### PR TITLE
[DNM] jewel: rgw: multisite: radosgw-admin period commands cant use --remote in another zonegroup

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/20029

### CODE REVIEW

**Comment by smithfarm** [comment]
```
rgw/rgw_rest_conn.cc: In constructor ‘RGWRESTConn::RGWRESTConn(RGWRESTConn&&)’:
rgw/rgw_rest_conn.cc:26:27: error: ‘class ceph::atomic_t’ has no member named ‘load’
     counter(other.counter.load())
                           ^
rgw/rgw_rest_conn.cc: In member function ‘RGWRESTConn& RGWRESTConn::operator=(RGWRESTConn&&)’:
rgw/rgw_rest_conn.cc:37:27: error: ‘class ceph::atomic_t’ has no member named ‘load’
   counter = other.counter.load();
                           ^
Makefile:24812: recipe for target 'rgw/librgw_la-rgw_rest_conn.lo' failed
make[3]: *** [rgw/librgw_la-rgw_rest_conn.lo] Error 1
```

**Comment by cbodley** [comment]
@smithfarm thanks. these conflicts are related to the change from `atomic_t` to `std::atomic<>` (prs https://github.com/ceph/ceph/pull/14839 and https://github.com/ceph/ceph/pull/15001). to resolve, change `.load()` to `.read()`

**Comment by smithfarm** [comment]
I guess `counter` is of type `ceph::atomic_t` and that type has an overloaded `=` operator whose method is not public?

```
In file included from ./common/ceph_context.h:25:0,
                 from ./global/global_context.h:18,
                 from ./common/dout.h:19,
                 from ./include/Context.h:19,
                 from rgw/rgw_rados.h:10,
                 from rgw/rgw_rest_conn.cc:4:
./include/atomic.h: In member function ‘RGWRESTConn& RGWRESTConn::operator=(RGWRESTConn&&)’:
./include/atomic.h:134:15: error: ‘ceph::atomic_t& ceph::atomic_t::operator=(const ceph::atomic_t&)’ is private
     atomic_t &operator=(const atomic_t &rhs);
               ^
rgw/rgw_rest_conn.cc:37:11: error: within this context
   counter = other.counter.read();
           ^
Makefile:24812: recipe for target 'rgw/librgw_la-rgw_rest_conn.lo' failed
```

**Comment by smithfarm** [comment]
Oh, now I see the problem. The `RGWRESTConn` class has an element `counter` which is of type `atomic_t`. This element cannot be assigned to - it's forbidden to use the `=` operator on it.

**Comment by smithfarm** [comment]
@cbodley Is that last commit OK?

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -1363,32 +1363,46 @@ int do_check_object_locator(const string& tenant_name, const string& bucket_name
   return 0;
 }
 
-#define MAX_REST_RESPONSE (128 * 1024) // we expect a very small response
-static int send_to_remote_gateway(const string& remote, req_info& info,
-                                  bufferlist& in_data, JSONParser& parser)
+/// search for a matching zone/zonegroup id and return a connection if found
+static boost::optional<RGWRESTConn> get_remote_conn(RGWRados *store,
+                                                    const RGWZoneGroup& zonegroup,
+                                                    const std::string& remote)
```
Comment (smithfarm) [inline src/rgw/rgw_admin.cc]: This causes a build failure on CentOS 7.3 and Ubuntu Trusty, which presumably have too-old boost versions:

```
rgw/rgw_admin.cc: In function 'boost::optional<RGWRESTConn> get_remote_conn(RGWRados*, const RGWZoneGroup&, const string&)':
rgw/rgw_admin.cc:1373:10: error: 'class boost::optional<RGWRESTConn>' has no member named 'emplace'
     conn.emplace(store->ctx(), store, remote, zonegroup.endpoints);
          ^
```

**Comment by smithfarm** [comment]
@cbodley This (apparently) needs to be reworked to get rid of `boost::optional` because (apparently) it's not in whatever version of Boost that ships with CentOS 7.3 and Ubuntu Trusty.

**Comment by cbodley** [comment]
@smithfarm okay thanks - `boost::optional` is there, but not its member function `emplace()`. i replaced that call with an assignment and pushed a copy of this branch to https://github.com/cbodley/ceph/commits/wip-pr-jewel-15458. i squashed this change into the `radosgw-admin: get remote connections from given period/zonegroup` commit and noted the conflict

**Comment by smithfarm** [comment]
@cbodley Thanks, much appreciated.

**Comment by smithfarm** [comment]
Build failure

**Comment by cbodley** [comment]
@smithfarm oops, i'll take a look

**Comment by cbodley** [comment]
@smithfarm i updated that same commit on branch https://github.com/cbodley/ceph/commits/wip-pr-jewel-15458, and actually verified that it compiles this time (sorry, jewel builds take forever without cmake!)

**Comment by smithfarm** [comment]
OK, re-updated and re-pushed.

**Comment by smithfarm** [comment]
Pushed `wip-20029-jewel` to ceph-ci: https://shaman.ceph.com/builds/ceph/wip-20029-jewel/f6699ca12741b81bd95e484ac9991b0ea91e448f/

**Comment by cbodley** [comment]
@smithfarm sorry, i haven't forgotten this one. i'm having trouble finding a way to do it correctly with the older boost versions

**Comment by smithfarm** [comment]
jenkins re-test this please

**Comment by smithfarm** [comment]
Jewel is EOL

**Pull request closed without merging**